tests/server.c

```diff
@@ -0,0 +1,368 @@
+      for(;;) {
+        result = do_read(conn, rconn);
+        if(result == CRUSTLS_DEMO_AGAIN) {
+          break;
+        }
+        else if(result == CRUSTLS_DEMO_CLOSE_NOTIFY) {
+          goto cleanup;
+        }
+        else if(result != CRUSTLS_DEMO_OK) {
+          goto cleanup;
+        }
+      }
+    }
+    if(FD_ISSET(sockfd, &write_fds)) {
+      fprintf(stderr, "rustls wants us to write_tls.\n");
+      err = rustls_connection_write_tls(rconn, write_cb, conn, &n);
+      if(err != 0) {
+        fprintf(stderr, "Error in write_tls: errno %d\n", err);
+        goto cleanup;
+      }
+      else if(n == 0) {
+        fprintf(stderr, "EOF from write_tls\n");
+        goto cleanup;
+      }
+    }
+
+    if(state == READING_REQUEST && body_beginning(&conn->data) != NULL) {
+      state = SENT_RESPONSE;
+      fprintf(stderr, "writing response\n");
+      if(send_response(conn) != CRUSTLS_DEMO_OK) {
+        goto cleanup;
+      };
+    }
+  }
+
+  fprintf(stderr, "handle_conn: loop fell through");
+
+cleanup:
+  if(sockfd > 0) {
+    close(sockfd);
+  }
+}
+
+const struct rustls_certified_key *
+load_cert_and_key(const char *certfile, const char *keyfile)
+{
+  char certbuf[10000];
+  size_t certbuf_len;
+  char keybuf[10000];
+  size_t keybuf_len;
+
+  int result = read_file(certfile, certbuf, sizeof(certbuf), &certbuf_len);
+  if(result != CRUSTLS_DEMO_OK) {
+    return NULL;
+  }
+
+  result = read_file(keyfile, keybuf, sizeof(keybuf), &keybuf_len);
+  if(result != CRUSTLS_DEMO_OK) {
+    return NULL;
+  }
+
+  const struct rustls_certified_key *certified_key;
+  result = rustls_certified_key_build((uint8_t *)certbuf,
+                                      certbuf_len,
+                                      (uint8_t *)keybuf,
+                                      keybuf_len,
+                                      &certified_key);
+  if(result != RUSTLS_RESULT_OK) {
+    print_error("parsing certificate and key", result);
+    return NULL;
+  }
+  return certified_key;
+}
+
+int
+main(int argc, const char **argv)
+{
+  int ret = 1;
+  int result = 1;
+  struct rustls_server_config_builder *config_builder =
+    rustls_server_config_builder_new();
+  const struct rustls_server_config *server_config = NULL;
+  struct rustls_connection *rconn = NULL;
+
+  if(argc <= 2) {
+    fprintf(stderr,
+            "usage: %s cert.pem key.pem\n\n"
+            "Listen on port 8443 with the given cert and key.\n",
+            argv[0]);
+    goto cleanup;
+  }
+
+  const struct rustls_certified_key *certified_key =
+    load_cert_and_key(argv[1], argv[2]);
+  if(certified_key == NULL) {
+    goto cleanup;
+  }
+
+  rustls_server_config_builder_set_certified_keys(
+    config_builder, &certified_key, 1);
+  server_config = rustls_server_config_builder_build(config_builder);
+
+#ifdef _WIN32
+  WSADATA wsa;
+  WSAStartup(MAKEWORD(1, 1), &wsa);
+#endif
+
+  int sockfd = socket(AF_INET, SOCK_STREAM, 0);
```

Comment:
`sockfd` is never closed, but it's probably fine to leak it since it'll get closed no matter what when the program exits.